api/v1alpha7/types.go

```diff
@@ -163,6 +163,15 @@ type RootVolume struct {
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
 }
 
+type DataVolume struct {
+	NameSuffix       string `json:"nameSuffix"`
+	Size             int    `json:"diskSize,omitempty"`
+	VolumeType       string `json:"volumeType,omitempty"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
```

Comment:
@stephenfin Very much interested in your input here.

I've proposed adding `useMachineAZ` (default false) to a *future* api version. What do you think the semantics of this one should be?

Options:
* Empty AZ means no AZ
* Empty AZ means something machine related
* Anything else?

Reply:
I was planning to make the semantics the same as the root volume for now. I think that is less confusing.

Then when we change the root volume we can change this as well. My plan is to have one "volume creating" function so hopefully the logic doesn't have to be duplicated.

Reply:
The problem is we know the semantics of the rootVolume are wrong, and we're going to have a minor upgrade problem there. If we know what we're going to do with rootVolume I'd prefer to get this right in the first instance.

See https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1662 for WIP.

Reply:
What is left to do on the root volume WIP? Should we get that merged first?

Reply:
I guess the issue here is you don't want to fall foul of the cross_az_attach checks, for clouds where that matters. In theory the Cinder AZ is independent from the Nova AZ, but most people don't do that. Network AZ on the ports is different again, of course.

FWIW, I think the main three cases are:
- 80% case: don't specify AZ at all, for anything, nova, neutron and cinder
- Volume and server AZ matches, to pass the cross_az_attach check (mostly for when you have multiple fibrechannel storage appliances, or similar, I think)
- Allow people to specify a custom AZ for storage that can be unrelated to the Nova and network AZ (because people bought multiple storage appliances, and want to expose that to their users for some reason, shrug)

(I am curious if there are other cases people care about.)

Reply:
@mkjpryor We need a firm decision on this, btw.

Reply:
I'd rather we did the same thing as the root volume in this PR, then change them all at once in another PR, personally, so that it stays consistent.

My preference would be that we add a flag that says `useServerAZ` and defaults to `false`, despite that being a breaking change, and keep the existing `availabilityZone` override. The logic would be:

1. If `availabilityZone` is set use it
2. If `useServerAZ = true`, set the AZ to the server AZ
3. Don't set an AZ at all

Reply:
I guess if we want a non-breaking change, then the `useServerAZ` flag defaults to `true`.

Reply:
> I guess if we want a non-breaking change, then the useServerAZ flag defaults to true.

That's what I believe is already CAPO's behavior on rootVolumes:

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/9d183bd479fe9aed4f6e7ac3d5eee46681c518e7/pkg/cloud/services/compute/instance.go#L439-L442

Reply:
Yes - I mirrored the current root behaviour for the additional volumes on purpose.

Reply:
We came to an agreement to bulk change it via another PR.